aten/src/ATen/native/TensorIndexing.cpp

```diff
@@ -154,5 +39,174 @@ std::ostream& operator<<(std::ostream& stream, const std::vector<TensorIndex>& t
+    }
+  }
+
+  std::vector<Tensor> tensorIndices;
+  Tensor sliced = applySlicing(self, indices, tensorIndices, self_device, self_sizes);
+  if (tensorIndices.empty()) {
+    if (sliced.is_same(self)) {
+      // ensure we return a shallow copy for things like x[...]
+      sliced = at::alias(sliced);
+    }
+    return sliced;
+  }
+
+  // indexing by tensors ("advanced" indexing)
+  return dispatch_index(sliced, std::move(tensorIndices));
+}
+
+// This mirrors `THPVariable_setitem` in torch/csrc/autograd/python_variable_indexing.cpp
+// for "the assigned value is a Tensor" case
+void set_item(Tensor& self, ArrayRef<TensorIndex> indices, const Tensor& value) {
+  OptionalDeviceGuard device_guard(device_of(self));
+  const at::Device self_device = self.device();
+  const IntArrayRef& self_sizes = self.sizes();
+
+  // handle simple types: integers, slices, ellipsis, bool
+  if (indices.size() == 1) {
+    const TensorIndex& index = indices[0];
+    if (!index.is_tensor()) {
+      return handleSimpleTypesInSingleDimIndexingSet(self, index, value, false, self_device, self_sizes);
+    }
+  }
+
+  std::vector<Tensor> tensorIndices;
+  Tensor sliced = applySlicing(self, indices, tensorIndices, self_device, self_sizes);
+  if (tensorIndices.empty()) {
+    copy_to(sliced, value);
+    return;
+  }
+
+  IntArrayRef valueSizes = value.sizes();
+  IntArrayRef slicedValueSizes = slicePrefix1sSize(valueSizes);
+  Tensor valuesSliced;
+  if (!valueSizes.equals(slicedValueSizes)) {
+    valuesSliced = value.view(slicedValueSizes);
+  } else {
+    valuesSliced = value;
+  }
+  dispatch_index_put_(sliced, std::move(tensorIndices), valuesSliced);
+  return;
+}
+
+// This mirrors `THPVariable_setitem` in torch/csrc/autograd/python_variable_indexing.cpp
+// for "the assigned value is a Scalar" case
+void set_item(Tensor& self, ArrayRef<TensorIndex> indices, Scalar v) {
+  OptionalDeviceGuard device_guard(device_of(self));
+  Tensor value;
+
+  {
+    at::AutoNonVariableTypeMode guard;
+    // TODO: This qint special case looks very suspicious...
+    if (isQIntType(self.scalar_type())) {
+      value = at::indexing::scalarToTensor(v, device(kCPU).dtype(kFloat), at::Device(kCPU));
+    } else {
+      value = at::indexing::scalarToTensor(v, self.options(), self.device());
+    }
+  }
+
+  return set_item(self, indices, value);
+}
+
 } // namespace indexing
+
+Tensor Tensor::index(ArrayRef<at::indexing::TensorIndex> indices) const {
+  return at::indexing::get_item(*this, indices);
+}
+Tensor Tensor::index(std::initializer_list<at::indexing::TensorIndex> indices) const {
+  return index(ArrayRef<at::indexing::TensorIndex>(indices));
+}
+
+Tensor & Tensor::index_put_(ArrayRef<at::indexing::TensorIndex> indices, Tensor const & rhs) {
+  at::indexing::set_item(*this, indices, rhs);
+  return *this;
+}
+Tensor & Tensor::index_put_(ArrayRef<at::indexing::TensorIndex> indices, Tensor && rhs) {
+  at::indexing::set_item(*this, indices, rhs);
```

Comment:
You need a `std::move(rhs)` here or else a copy (refcount bump) will happen.

Reply:
...I say, but I see that `set_item` only takes Tensor by const reference. If this is the case, why do you need an explicit rvalue overload at all in the first place? `const Tensor&` will work just fine for temporaries.